[Merged by Bors] - P2P decentralization improvements

ivan4th opened the pull request:
## Motivation

This PR implements changes needed for spacemeshos/pm#275, except for measurement

## Changes
* Introduce Routing Discovery to contact peers behind NATs
* Introduce dynamic v2 relay discovery which is needed for hole punching. The idea is to have a wider array of circuit-v2 passive relays which should be much safer than old libp2p active relays (which were disabled in e.g. Filecoin due to security concerns)
* Introduce QUIC transport to improve chances at hole punching, with testnet-mainnet "crosstalk" protection based on a transport-level handshake mechanism
  * the handshake is not used on mainnet. That way, connections between mainnet and testnet nodes are still prevented, as testnet peers expect the handshake, but if/when my libp2p changes are merged (libp2p/go-libp2p#2658) or libp2p gets private network support
* Make it possible to listen on multiple addresses and advertise multiple addresses
* Extend DebugService with additional P2P info needed for hole punching diagnostics (needs spacemeshos/api#285)
* Add `ping-peers` config option to facilitate P2P network issue diagnostics
* Add `force-dht-server` config option that is useful during troubleshooting DHT and hole-punching issues

`ping-peers` and `force-dht-server` were initially considered to be temporary features, but I think it might make sense to keep them for various P2P network troubleshooting scenarios.

All of the changes are disabled in the config by default, except for:
* libp2p Ping service is enabled by default to make diagnostics easier
* DHT Values and Providers as these will make DHT Routing Peer discovery work efficiently from the beginning when we enable this feature in the configs
* Bootnodes aren't used as relays by default anymore. v2 relays have very limited capacity by default and bootnode relay servers' reservations are very quickly exhausted. Need to either specify a static relay list or enable routing discovery, which searches for more available relays as needed

## Test Plan
* Tested using k8s several clusters with cone NATs enabled via `bridge` CNI plugin (via Multus) -- backported to v1.2.8
* Added a Mac node for testing

## TODO
- [x] Have spacemeshos/api#285 merged and updated to the new `api` release
- [x] Retest using an image based on this branch (not backport)
- [ ] Decide on whether/how to extend systests to include NAT testing
- [ ] To check: TCP holepunching tends to happen more than QUIC (might be related to the handshake mechanism)
- [ ] ~~To consider: try picking up some % (e.g.: 50%) of non-infra peers during routing discovery~~ (doesn't work too well, need something more involved for that)

Maybe as a follow-up (depending on how soon this gets reviewed):
- Include new metrics / check if they're already present
  - NAT type (UDP / TCP) - Cone / Symmetric / Unknown
  - Reachability - Public / Private / Unknown
  - N of "advertised" peers found via routing discovery
  - N of TCP and UDP (QUIC) peers
  - N of peers reached via relayed connections (these being present for a long time may indicate hole-punching troubles, usually relayed connections go away relatively quickly)
  - N of relay reservations this node managed to obtain
  - Whether routing discovery is active or suspended (e.g. b/c `low-peers` N of peers has been reached)
  - Whether DHT is in the `Server` or `Client` mode
- systests checking NATed connections

ivan4th committed:
p2p: enable QUIC support

ivan4th committed:
p2p: add advertise-addresses and ping-peers

ivan4th committed:
p2p: add force-dht-server option

ivan4th committed:
p2p: add listen-addresses config option

ivan4th committed:
p2p: add genesis time validation for QUIC connections

ivan4th committed:
p2p: fix broken test, compat and linter errors

ivan4th committed:
p2p: fix listen and advertise-address command line options

Allow multiple values

ivan4th committed:
p2p: fixup: listen-addresses / advertise-addresses options

ivan4th committed:
p2p: make it possible to enable/disable transports from the config

ivan4th committed:
p2p: use libp2p routing discovery

ivan4th committed:
p2p: use transport-level handshake

ivan4th committed:
p2p: use ping service for ping-peers

ivan4th committed:
p2p: do not use QUIC NetworkCookie on the mainnet

ivan4th committed:
p2p: provide more P2P info through DebugService

ivan4th committed:
p2p: improve ping and routing discovery

Suspend routing discovery when there's enough peers
Use DHT to find ping peers

ivan4th committed:
p2p: add missing command line flags

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,160 @@
+				select {
+				case <-p.ctx.Done():
+					cancel()
+					return p.ctx.Err()
```
The call to `cancel` here shouldn't be necessary if its parent context was already completed.

```suggestion
				case <-p.ctx.Done():
					return p.ctx.Err()
```

ivan4th replied:
Added due to linter complaints. Perhaps need a comment for that (or disable the linter warning in this place)

ivan4th replied:
Ping code cleaned up

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,160 @@
+
+type Ping struct {
+	sync.Mutex
+	logger log.Log
```
We are trying to deprecate our `log` package, if possible use `*zap.Logger` directly.

ivan4th replied:
ok, will switch to `zap`

ivan4th replied:
Ping code now uses `zap`

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,160 @@
+	peers  []peer.ID
+	pr     routing.PeerRouting
+	stats  map[peer.ID]*pingStat
+	ctx    context.Context
```
`struct`s should not contain a `context.Context`: https://go.dev/blog/context-and-structs

Instead you could pass the context in `(*Ping) Start` directly to `p.startForPeer` and only keep `cancel` for `(*Ping) Stop`.

ivan4th replied:
Yes, makes sense, I'm aware of this guideline and will fix it.
Was a bit confused by some other parts of the code though
https://github.com/spacemeshos/go-spacemesh/blob/3d8eedfd44c6f9f8987e604b768596f6001fe703/hare3/hare.go#L190

ivan4th replied:
Ping code cleaned up

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,160 @@
+	return ping.Ping(ctx, p.h, peerID), nil
+}
+
+func (p *Ping) startForPeer(peerID peer.ID) {
+	p.h.ConnManager().Protect(peerID, "sm-ping")
+
+	p.eg.Go(func() error {
+		var ctx context.Context
+		var cancel context.CancelFunc
+		defer func() {
+			if cancel != nil {
+				cancel()
+			}
+		}()
+	OUTER:
+		for {
+			if cancel != nil {
+				cancel()
+			}
+			ctx, cancel = context.WithCancel(p.ctx)
+			ch, err := p.doPing(ctx, peerID)
+			if err != nil {
+				cancel()
+				p.logger.With().Error("pre-PING connect failed", log.Err(err))
+				select {
+				case <-p.ctx.Done():
+					return p.ctx.Err()
+				case <-time.After(pingInterval):
+					continue
+				}
+			}
```
A `context.CancelFunc` can be called any amount of times (all but the first call are no-ops) and calling `cancel` of a child context isn't necessary when the parent context was already canceled. The code could be shortened to:

```suggestion
func (p *Ping) startForPeer(ctx context.Context, peerID peer.ID) {
	p.h.ConnManager().Protect(peerID, "sm-ping")

	p.eg.Go(func() error {
	OUTER:
		for {
			pingCtx, cancel := context.WithCancel(ctx)
			defer cancel() // this servers to avoid a dead go routine if we leave the scope and forget calling cancel explicitly
			ch, err := p.doPing(pingCtx, peerID)
			if err != nil {
				cancel()
				p.logger.With().Error("pre-PING connect failed", log.Err(err))
				select {
				case <-ctx.Done():
					return ctx.Err()
				case <-time.After(pingInterval):
					continue
				}
			}
```

ivan4th replied:
Some extra cancels were added due to linter complaints about missing cancels on some of the code paths

ivan4th replied:
Ping code cleaned up

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,160 @@
+				}
+			}
+
+			cancel()
```
This can be dropped if the call to `cancel` was deferred immediately after creating it (recommended to not accidentally create dead go routines).

ivan4th replied:
There were linter complaints about missing `cancel()` calls

fasmat replied:
Are you using `golang-ci-lint` or your IDEs internal linter? Normally when you defer `cancel` right after instantiating it, the linter shouldn't complain 🙂

ivan4th replied:
I was doing `make lint` which uses `golang-ci-lint`. Will recheck

ivan4th replied:
Ping code cleaned up

fasmat commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/spacemeshos/go-spacemesh
 
 go 1.21.3
 
+replace github.com/libp2p/go-libp2p => github.com/ivan4th/go-libp2p v0.0.0-20231115132946-8d742f4b0068
```
How likely will your changes be accepted by `libp2p`? We should avoid having to maintain a fork over a longer time period.

ivan4th replied:
Sorry, my mistake, this fork is no longer needed b/c I implemented QUIC handshake on our side. Perhaps messed up a bit when shuffling around the commits between 1.2.8 and develop-based test branches. Will fix that

ivan4th replied:
Removed the replacement

ivan4th committed:
p2p: log resuming routing discovery

ivan4th committed:
p2p: in case of an unknown NAT type, still try routing discovery

ivan4th committed:
p2p: further improve routing discovery

Make advertising for peer discovery optional
Stop routing discovery after lowPeers
Refactor
Improve test

ivan4th committed:
p2p/dhtdiscovery: refactor, improve test, fix corner case issues

ivan4th committed:
p2p ping: refactor, use zap for logging

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,158 @@
+	return p
+}
+
+func (p *Ping) Start(ctx context.Context) {
```
This context is unused.

```suggestion
func (p *Ping) Start() {
```

ivan4th replied:
I've used `context.Background()` instead of `ctx` by mistake below, will fix, thanks

fasmat replied:
`context.Background()` is correct otherwise you have 2 ways to stop the ping: `Stop` and canceling the context passed to start 🙂

ivan4th replied:
Fixed

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,158 @@
+				zap.Error(err))
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
```
If you return an error here you disable `p.eg` from accepting new go routines, i.e. a `(*Ping) Start()` -> `(*Ping) Stop()` -> `(*Ping) Start()` might not work if `p.runForPeer` returns an error and `Stop` is called during the `pingInterval`.

ivan4th replied:
Fixed

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,158 @@
+// runForPeer runs ping for the specific peer till an error occurs
+// or the context is canceled.
+func (p *Ping) runForPeer(ctx context.Context, peerID peer.ID, addrs []ma.Multiaddr) error {
+	pingCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ch, err := p.doPing(pingCtx, peerID)
```
I think you can just pass the parent context here:

```suggestion
	ch, err := p.doPing(ctx, peerID)
```

ivan4th replied:
If there's an error sent on the channel returned by libp2p's `Ping`, it appears that it stops working, most of the time, although this behavior is not documented and probably not intended, and may change in future versions. So what I want here is to always stop and restart ping when I get an error from the channel, thus I create a child context and cancel it when leaving `doPing`, which is then called again, after a delay, if there was a ping error.
I will add a comment to clarify that.

fasmat commented on p2p/ping.go:
```diff
@@ -0,0 +1,158 @@
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
+			p.logger.Error("PING failed",
+				zap.Stringer("peer", peerID),
+				zap.Any("addrs", addrs),
+				zap.Error(err))
```
Is error the correct logging level here, i.e. is something wrong with the node if this line appears in the log?

ivan4th replied:
`Ping` is a network troubleshooting feature which is intended to be only used for diagnostic purposes, namely, to understand connectivity issues between nodes which must be able to talk to each other but somehow can't. When it is enabled, it's preferable to give ping messages higher visibility even if logging level is reduced, thus I use `Error` for ping errors and `Info` for ping successes.

fasmat commented on p2p/host.go:
```diff
@@ -90,6 +99,7 @@ type Config struct {
 	DisableDHT                  bool        `mapstructure:"disable-dht"`
 	Flood                       bool        `mapstructure:"flood"`
 	Listen                      string      `mapstructure:"listen"`
+	ListenAddresses             []string    `mapstructure:"listen-addresses"`
```
What happens to the old config values like `Listen` and `AdvertiseAddress`? They seem to overwrite the new config fields if both are set.

Could you add a warning if a user does this? Ideally with a hint that this is deprecated and will be dropped in a future version? (also a ticket to actually drop the old fields from the config & CLI would be good 🙂  )

ivan4th replied:
Yes, I need to make config validation fail if this happens. Will fix

ivan4th replied:
Upon further consideration, found out (or, to be more precise, remembered) that the problem cannot be fixed that easily due to backward compatibility concerns. Namely, previously we had ListenAddress (`listen-address`) which was just a string, and now we have ListenAddresses (`listen-addresses`) which is a list of strings. The value must default to `/ip4/0.0.0.0/tcp/7513`. My idea was to produce an error when both `listen-address` and `listen-addresses` are set, but the problem is, if one of these fields has a default value, the validation will always fail when the other is specified. So, I had to resort to just ignoring `listen-addresses` when `listen-address` is specified, which is far from perfect, but at least is backward-compatible. So we either need to keep this scheme or alter the defaulting mechanism such as having no default for the config field and using default `ListenAddresses` only when configuring the host.

Re `advertise-addresses`, we could handle it "better", but that would be somewhat inconsistent

fasmat replied:
We want to force users to update their config to the new fields and print an error during startup if they still use an old (and now invalid) config and not keep code around that tries to make old versions of a config file somehow work with newer versions of the node - to avoid overly complicating our own config parsing over time.

To achieve this you can do the following:

- specify a custom type `DeprecatedListener` (or similar) that is a `string` but also implements the `pflag.Value` interface
- in the `Set` method of this interface you can return an `error` and you need to update `EnsureCLIFlags` in `base.go` with an additional case with your new type that also panics when the value is tried to be set
- for examples of custom types in our config you can see `activation.PowDifficulty` or `activation.PostProviderID`

The error that you return or the panic you throw should indicate to the user that the config field is not valid any more and that they have to specify the new field instead. Also add that information to `CHANGELOG.md`.

This way users are forced to update their config to the new schema. We can remove the old (now invalid/unused fields) in a future version where they will then just be ignored.

ivan4th replied:
I like this idea but what I'm afraid here is impact on our own infra, in particular. This might have to be adjusted
in multiple places and that might cause some extra hassle; I will ask other folks about their opinion on this, too.

But that actually gave me a possible idea how to deal with it in a backward-compatible way:
* define a new type that is `[]string` and can be decoded from both a list of strings and just a string by mapstructure, e.g. `type StringList []string`
* instead of adding new fields, just make `Listen` and `AdvertiseAddress` fields to be `StringList`. This way, old configs will still be handled, and `--listen` and `--advertise-address` options will still work, and you can also pass multiple `--listen` and `--advertise-address` options to specify multiple addrs.

We can optionally log warnings about `StringList` being decoded from simple string in the config, and later remove this possibility, should that be needed.

The drawback of this approach is slight misnomer, b/c in the config we'll have `advertise-address` which can be a list of strings instead of just a single string. For `listen` I think it is ok.

ivan4th replied:
Implemented in backward-compatible way like I described above, but using
```golang
type AddressList []multiaddr.Multiaddr
```

fasmat commented on p2p/host.go:
```diff
@@ -111,6 +123,14 @@ type Config struct {
+	EnableQUICTransport         bool        `mapstructure:"enable-quic-transport"`
+	EnableRoutingDiscovery      bool        `mapstructure:"enable-routing-discovery"`
+	RoutingDiscoveryNoAdvertise bool        `mapstructure:"routing-discovery-no-advertise"`
+	// TBD: adv peer fraction
```
Will this still be added? If not it would be good to have a ticket for it and reference it here.

ivan4th replied:
That was a failed experiment of forcing the node to have no less than some % of advertised (non-infra) peer nodes and I forgot to remove the comment, sorry. (the subject of the experiment proved to be unnecessary)
Will fix

ivan4th replied:
Fixed

fasmat commented on p2p/host.go:
```diff
@@ -198,63 +238,98 @@ func New(
+		for _, addrStr := range advAddrs {
+			addr, err := multiaddr.NewMultiaddr(addrStr)
+			if err != nil {
+				panic(err) // validated in config
```
I'm not sure if this is actually true; at least I can find any place where `AdvertiseAddress` or `AdvertiseAddresses` are actually validated to be valid multiaddresses.

ivan4th replied:
The code that validates the addresses is here: https://github.com/spacemeshos/go-spacemesh/blob/a1a2e40d78b014d39536663e47f0641d03c3d5bb/p2p/host.go#L178-L183
But it had a bug in it (listen addresses skipped by mistake), so thanks for noticing :)

fasmat replied:
Similarly to my other comment you can make `ListenAddresses` and `AdvertiseAdresses` custom types that extend `ma.Multiaddress` but also implement the `pflag.Value` interface and have the validation happen when the config/CLI is parsed

This way you already have a valid slice of multiadresses here and if the config is invalid it already fails to start. 🙂

fasmat replied:
You can also take a look at `postproviderid.go` in the `mapstructureutil` package for an example on how to decode the config into a custom type.

ivan4th replied:
Fixed in a backward-compatible way to reduce devops hassles, see above.

fasmat commented on p2p/host.go:
```diff
@@ -111,6 +123,14 @@ type Config struct {
 	GossipValidationThrottle    int         `mapstructure:"gossip-validation-throttle"`
 	GossipAtxValidationThrottle int         `mapstructure:"gossip-atx-validation-throttle"`
+	PingPeers                   []string    `mapstructure:"ping-peers"`
+	Relay                       bool        `mapstructure:"relay"`
+	Relays                      []string    `mapstructure:"relays"`
```
Would it make sense to only have `Relays`? If at least one is defined `Relay` is `true`, otherwise it is `false`, no need for a separate config 🙂

ivan4th replied:
`Relays` need to be renamed to `StaticRelays`. That's an alternative possibility of specifying a relay list instead of enabling relay discovery, and that might be useful during network troubleshooting. But naming is confusing here indeed, will fix

ivan4th replied:
(if static relays aren't specified, routing discovery method is used to find available relays)

ivan4th replied:
Fixed

fasmat commented on p2p/host.go:
```diff
@@ -198,63 +238,98 @@ func New(
 		resources.ReservationTTL = cfg.RelayServer.TTL
 		lopts = append(lopts, libp2p.EnableRelayService(relay.WithResources(resources)))
 	}
+	if cfg.Relay {
+		lopts = append(lopts, libp2p.EnableRelay())
+	}
+	if len(cfg.Relays) != 0 {
+		relays, err := parseIntoAddr(cfg.Relays)
+		if err != nil {
+			return nil, err
+		}
+		lopts = append(lopts, libp2p.EnableAutoRelayWithStaticRelays(relays))
+	} else {
+		peerSrc, relayCh := relayPeerSource(logger)
+		lopts = append(lopts, libp2p.EnableAutoRelayWithPeerSource(peerSrc))
+		opts = append(opts, WithRelayCandidateChannel(relayCh))
+	}
```
I'm not fully understanding how these options work together; doesn't `cfg.Relay` always need to be true when `cfg.Relays` contains at least one relay?

ivan4th replied:
This indeed needs to be fixed. According to the libp2p API, the `EnableRelay()` option seems to be redundant, as the relay is enabled by default if `DisableRelay()` is not called: https://github.com/libp2p/go-libp2p/blob/4b6955b7dc992882a237b66c7f8577eb8293dabc/options.go#L271-L282
But the comment contains outdated information at least in the part that states that v1 relays are enabled, as support for v1 active relay has been dropped from the `go-libp2p` codebase, so I'm unsure about the relay enablement too, and will recheck. And yes, if the relay is not enabled then the options related to relay discovery / static lists don't make sense, I will fix that

ivan4th replied:
Fixed, also made the code explicitly disable the relays when `Relay: false` is used

fasmat commented on p2p/upgrade.go:
```diff
@@ -217,19 +349,29 @@ func (fh *Host) PeerProtocols(p Peer) ([]protocol.ID, error) {
+	fh.discovery.Start(fh.ctx)
+	if fh.ping != nil {
+		fh.ping.Start(fh.ctx)
+	}
```
As I noted in a previous comment passing `context.Context` to Start is confusing, as it is not clear what canceling the context does: is it analogous to `Stop` or does it cancel the `Start`? Why is `Start` even possible if `fh.ctx` is already canceled?

(In the concrete case here `Stop` still has to be called explicitly or `fh.discovery` will not be closed correctly)

It is out of scope to remove the context from the `Host` struct, but in my opinion it should not be passed to sub-routines here and in a future change removed entirely from the object.

ivan4th replied:
Removed `ctx` from Start arguments, `Stop` needs to be called to stop the discovery

fasmat commented on p2p/dhtdiscovery/discovery_test.go:
```diff
@@ -49,5 +130,27 @@ func TestSanity(t *testing.T) {
+		if ch == nil {
+			continue
+		}
+		wg.Add(1)
+		go func() {
+			select {
+			case p := <-ch:
+				require.Equal(t, mock.Hosts()[1].ID(), p.ID)
+			case <-time.After(3 * time.Second):
+				t.Errorf("timed out waiting for relay id for peer %d", n)
+			}
+			wg.Done()
+		}()
```
You can use an `errgroup.Group` instead of a `sync.WaitGroup` here 🙂

ivan4th replied:
Well in this particular case, it doesn't matter much b/c the code in question doesn't make use of any special `errgroup` functionality and doesn't even return an `error` (just registers test failure instead), but I may indeed switch to `errgroup` here for the sake of consistency

ivan4th replied:
Fixed

fasmat commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -143,51 +178,70 @@ type Discovery struct {
+func (d *Discovery) setupDHT(ctx context.Context) error {
+	d.dhtLock.Lock()
+	defer d.dhtLock.Unlock()
+	return d.newDht(ctx, d.h, d.public, d.mode, d.dir)
```
Shall it be possible to call `newDht` without grabbing the `dhtLock` mutex? If not I would just inline the method here.

ivan4th replied:
Fixed

fasmat commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -199,20 +253,20 @@ func (d *Discovery) Stop() {
 	if err := d.dht.Bootstrap(ctx); err != nil {
 		d.logger.Error("unexpected error from discovery dht", zap.Error(err))
 	}
 	<-ctx.Done()
```
This `<-ctx.Done()` is basically a sleep for the remaining time of `d.bootstrapDuration` and shouldn't be needed?

ivan4th replied:
Well I didn't change this logic, but IIUC the idea here is this: we do indeed want this delay to have bootstrap run for some time before we check the number of peers and retry if it's too low. But if `Bootstrap()` call took some time by itself, we want to subtract that time from the delay, to keep the total bootstrap duration

fasmat commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +318,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
```
I think we should `defer` this call to `ticker.Stop` to ensure no matter how we exit this method the resources allocated by `time.NewTicker` are guaranteed to be freed.

ivan4th replied:
Fixed

ivan4th committed:
Address review comments

ivan4th committed:
Use backward-compatible listen/advertise-address options

ivan4th committed:
p2p: add more relay options

ivan4th committed:
p2p: make adverise interval configurable and default to 1 min

poszu commented on node/node_test.go:
```diff
@@ -800,6 +801,32 @@ func TestConfig_CustomTypes(t *testing.T) {
+			},
+		},
+		{
+			name: "address-list-single",
```
```suggestion
			name: "address-list-multiple",
```

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+		wasSuspended := false
+		for !d.h.NeedPeerDiscovery() {
+			wasSuspended = true
+			d.logger.With().Info("suspending routing discovery",
```
```suggestion
			d.logger.Info("suspending routing discovery",
```

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+			}
+		}
+		if wasSuspended {
+			d.logger.With().Info("resuming routing discovery")
```
```suggestion
			d.logger.Info("resuming routing discovery")
```

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+		if d.h.Network().Connectedness(p.ID) != network.Connected {
+			if _, err := d.h.Network().DialPeer(ctx, p.ID); err != nil {
+				d.logger.Error("error dialing peer", zap.Any("peer", p),
+					zap.Error(err))
```
Should this be an error-level log? Isn't it "normal" that sometimes dialing peer fails?

ivan4th replied:
Made it `Debug`

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+		}
+	}
+	return nil
+}
```
It never returns a non-nil error. How about changing to:
```suggestion
func (d *Discovery) discoverRelays(ctx context.Context) {
	for p := range d.findPeersContinuously(ctx, relayNS) {
		if len(p.Addrs) != 0 {
			d.logger.Debug("found relay candidate", zap.Any("p", p))
			select {
			case d.relayCh <- p:
			case <-ctx.Done():
				return
			}
		}
	}
}
```

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+			select {
+			case <-ctx.Done():
+				return nil
+			case p, ok := <-peerCh:
```
`peerCh` could be `nil` at this point (if `d.disc.FindPeers(ctx, ns)` failed). Reading from it would block forever then.

ivan4th replied:
Forgot to `continue` the loop in the `if` above, fixed, thanks

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+	return r
+}
+
+// TBD: don't store the context, create DHT during start
```
I think it's done already, isn't it?

ivan4th replied:
Yes, removed that stale comment, thanks

poszu commented on p2p/handshake/handshake.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+// WithLog specifies handshake retry count.
+func WithAttempts(n int) Option {
```
```suggestion
// WithAttempts specifies handshake retry count.
func WithAttempts(n int) Option {
```

poszu commented on p2p/ping.go:
```diff
@@ -0,0 +1,163 @@
+)
+
+const (
+	pingInterval   = time.Second
```
Would it make sense to have constants like this configurable to be able to tweak them without releasing a new version?

ivan4th replied:
Made ping interval configurable. It's a purely troubleshooting feature that I hope will not be needed _too often_, but indeed let's be on the safe side

ivan4th committed:
p2p: address review comments

ivan4th committed:
p2p: make ping interval configurable

ivan4th committed:
Update api to v1.25.0

This includes DebugService.NetworkInfo changes for P2P

ivan4th commented:
Fixed the conflict, created https://github.com/spacemeshos/api/releases/tag/v1.25.0 (also https://github.com/spacemeshos/api/releases/tag/release%2Fgo%2Fv1.25.0) `api` release and switched to it

ivan4th committed:
p2p: use bootnodes as relays when routing discovery is off

ivan4th committed:
p2p: do not do routing discovery advertisement by default

ivan4th committed:
p2p: do not discover relays when routing discovery is off

ivan4th committed:
Update CHANGELOG.md and README.md

ivan4th commented:
Disabled routing discovery advertisement by default, and made the nodes use bootnodes as relays when routing discovery is disabled for even better consistency with existing behaviour when an old config is being used.
Updated `README.md` and `CHANGELOG.md`

ivan4th commented:
@poszu pls check if the README / changelog changes are sufficient (or if they're excessive)

fasmat commented on p2p/upgrade.go:
```diff
@@ -182,7 +182,7 @@ func Upgrade(h host.Host, opts ...Opt) (*Host, error) {
 		dopts = append(dopts, discovery.EnableRoutingDiscovery())
 	}
-	if !fh.cfg.RoutingDiscoveryNoAdvertise {
+	if !fh.cfg.RoutingDiscoveryAdvertise {
```
Could it be that this should be

```suggestion
	if fh.cfg.RoutingDiscoveryAdvertise {
```

?

ivan4th replied:
Good catch! Broke this part in a last minute update... Fixed.

fasmat commented on p2p/handshake/handshake.go:
```diff
@@ -0,0 +1,278 @@
+	attempts      int
+}
+
+var (
+	_ transport.Transport = &transportWrapper{}
+	_ io.Closer           = &transportWrapper{}
+)
```
The first line somewhat redundant, since `transportWrapper` already embeds a `transport.Transport` so this type check will never fail 🙂 

Regarding `Close`: the transport that is wrapped does have a `Close` method, but it looks like this: https://github.com/libp2p/go-libp2p/blob/master/p2p/transport/quic/transport.go#L370-L372 If `transportWrapper` doesn't wrap any other transports, I don't think we need to make sure to call it.

ivan4th replied:
Concerning interface checks: well, actually there's a possibility for `transportWrapper` to fail to implement `transport.Transport`: using a wrong signature for an "overridden" method, such as `Listen()` or `Dial()`. I'd prefer to have a guard here against possible changes in `go-libp2p` method signatures (or our own mistakes).

Almost the same goes for `Close()` method: it being a nop is not really documented and for whatever reason may change in some future `go-libp2p` version, and calling this nop version doesn't do much harm, yet it keeps us from depending on libp2p implementation details.

fasmat commented on p2p/handshake/handshake.go:
```diff
@@ -0,0 +1,278 @@
+	attempts int
+}
+
+var _ transport.Listener = &listenerWrapper{}
```
same as above, this seems redundant when `listenWrapper` already is a `transport.Listener` via embedding

ivan4th replied:
Also, same as above: "overridden" methods can have wrong signature (e.g. after `go-libp2p` version update) and having these checks fail might save us a little bit of time when upgrading the deps someday in future

ivan4th commented:
bors merge
